File: sync/src/training_engine/banister.py
"""Banister impulse-response model for personalized PMC time constants.

Fits fitness (tau1) and fatigue (tau2) decay constants from actual run history
using maximal-effort anchor detection and differential evolution optimization.

The classic Banister model:
    p(t) = p0 + k1 * sum(w(i) * exp(-(t-i)/tau1)) - k2 * sum(w(i) * exp(-(t-i)/tau2))

Where:
    p0    = baseline performance (VDOT)
    k1    = fitness gain coefficient
    k2    = fatigue gain coefficient
    tau1  = fitness decay time constant (days, typically ~42)
    tau2  = fatigue decay time constant (days, typically ~7)
    w(i)  = training load on day i

References:
    - Banister, E.W. (1991). "Modeling elite athletic performance."
    - Busso, T. (2003). "Variable dose-response relationship between
      exercise training and performance."
"""

from __future__ import annotations

import json
import math
from dataclasses import dataclass
from typing import List, Tuple

from training_engine.vdot import vdot_from_race


@dataclass
class BanisterParams:
    """Fitted Banister model parameters."""
    p0: float     # baseline VDOT
    k1: float     # fitness gain coefficient
    k2: float     # fatigue gain coefficient
    tau1: float   # fitness decay time constant (days)
    tau2: float   # fatigue decay time constant (days)


# Default parameters when insufficient data for fitting
_DEFAULT_PARAMS = BanisterParams(p0=45.0, k1=0.05, k2=0.08, tau1=42, tau2=7)


def detect_anchor_runs(
    runs: list[dict],
    estimated_hrmax: float,
    hr_threshold_pct: float = 0.90,
    min_distance_m: float = 2000,
) -> list[dict]:
    """Detect maximal-effort anchor runs for Banister model calibration.

    Filters runs where avg_hr >= threshold % of HRmax AND distance >= min_distance.
    Computes VDOT for each qualifying run via the Daniels/Gilbert equation.

    Args:
        runs: List of run dicts with keys: date, avg_hr, distance_m, duration_s.
        estimated_hrmax: Estimated maximum heart rate (bpm).
        hr_threshold_pct: Minimum avg_hr / HRmax fraction (default 0.90).
        min_distance_m: Minimum distance in meters (default 2000).

    Returns:
        List of anchor dicts sorted by date, each containing the original
        run fields plus a computed 'vdot' value.
    """
    hr_cutoff = estimated_hrmax * hr_threshold_pct
    anchors = []

    for run in runs:
        avg_hr = run.get("avg_hr", 0)
        distance_m = run.get("distance_m", 0)
        duration_s = run.get("duration_s", 0)

        if avg_hr < hr_cutoff:
            continue
        if distance_m < min_distance_m:
            continue
        if duration_s <= 0:
            continue

        vdot = vdot_from_race(distance_m, duration_s)
        anchor = {**run, "vdot": vdot}
        anchors.append(anchor)

    anchors.sort(key=lambda a: a["date"])
    return anchors


def banister_predict(
    params: BanisterParams,
    daily_loads: list[tuple[int, float]],
    target_day: int,
) -> float:
    """Predict performance (VDOT) at a target day using the Banister model.

    p(t) = p0 + k1 * sum(w(i) * exp(-(t-i)/tau1)) - k2 * sum(w(i) * exp(-(t-i)/tau2))

    Args:
        params: BanisterParams with model coefficients.
        daily_loads: List of (day_index, load_value) tuples.
        target_day: Day index to predict performance at.

    Returns:
        Predicted VDOT score.
    """
    fitness_sum = 0.0
    fatigue_sum = 0.0

    for day_i, load in daily_loads:
        dt = target_day - day_i
        if dt <= 0:
            continue
        fitness_sum += load * math.exp(-dt / params.tau1)
        fatigue_sum += load * math.exp(-dt / params.tau2)

    return params.p0 + params.k1 * fitness_sum - params.k2 * fatigue_sum


def fit_banister(
    daily_loads: list[tuple[int, float]],
    anchors: list[dict],
    max_iterations: int = 1000,
) -> BanisterParams:
    """Fit Banister model parameters using differential evolution.

    Minimizes sum of squared errors between predicted and observed VDOT
    at anchor points. If fewer than 2 anchors are available, returns
    sensible defaults (tau1=42, tau2=7).

    Args:
        daily_loads: List of (day_index, load_value) tuples.
        anchors: List of dicts with 'day_index' and 'vdot' keys.
        max_iterations: Maximum iterations for differential evolution.

    Returns:
        Fitted BanisterParams.
    """
    if len(anchors) < 2:
        return BanisterParams(
            p0=_DEFAULT_PARAMS.p0,
            k1=_DEFAULT_PARAMS.k1,
            k2=_DEFAULT_PARAMS.k2,
            tau1=_DEFAULT_PARAMS.tau1,
            tau2=_DEFAULT_PARAMS.tau2,
        )

    from scipy.optimize import differential_evolution

    # Bounds: p0, k1, k2, tau1, tau2
    bounds = [
        (35, 55),       # p0: baseline VDOT
        (0.001, 5),     # k1: fitness gain
        (0.001, 10),    # k2: fatigue gain
        (20, 80),       # tau1: fitness decay (days)
        (3, 20),        # tau2: fatigue decay (days)
    ]

    def objective(x):
        p = BanisterParams(p0=x[0], k1=x[1], k2=x[2], tau1=x[3], tau2=x[4])
        sse = 0.0
        for anchor in anchors:
            predicted = banister_predict(p, daily_loads, anchor["day_index"])
            sse += (predicted - anchor["vdot"]) ** 2
        return sse

    result = differential_evolution(
        objective,
        bounds,
        maxiter=max_iterations,
        seed=42,
        tol=1e-8,
        polish=True,
    )

    return BanisterParams(
        p0=result.x[0],
        k1=result.x[1],
        k2=result.x[2],
        tau1=result.x[3],
        tau2=result.x[4],
    )


def load_anchors_from_db(conn, estimated_hrmax: float = 190) -> list[dict]:
    """Load anchor runs from garmin_activity_raw database table.

    Queries running activities from the 'summary' endpoint, extracts avg_hr,
    distance, and duration from the JSONB data, then filters through
    detect_anchor_runs().

    Args:
        conn: psycopg2 database connection.
        estimated_hrmax: Estimated maximum heart rate.

    Returns:
        List of anchor dicts with date, day_index, and vdot.
    """
    with conn.cursor() as cur:
        cur.execute("""
            SELECT activity_id, raw_json
            FROM garmin_activity_raw
            WHERE endpoint_name = 'summary'
              AND raw_json->'activityType'->>'typeKey' = 'running'
            ORDER BY (raw_json->>'startTimeLocal')::date ASC
        """)
        rows = cur.fetchall()

    runs = []
    for activity_id, raw_json in rows:
        data = raw_json if isinstance(raw_json, dict) else json.loads(raw_json)

        avg_hr = data.get("averageHR")
        distance_m = data.get("distance")
        duration_s = data.get("duration")
        start_local = data.get("startTimeLocal", "")

        if not all([avg_hr, distance_m, duration_s, start_local]):
            continue

        # Extract date string (first 10 chars of ISO timestamp)
        date_str = str(start_local)[:10]
        runs.append({
            "date": date_str,
            "avg_hr": float(avg_hr),
            "distance_m": float(distance_m),
            "duration_s": float(duration_s),
            "activity_id": activity_id,
        })

    return detect_anchor_runs(runs, estimated_hrmax=estimated_hrmax)


def _load_daily_loads_from_db(conn) -> tuple[list[tuple[int, float]], str]:
    """Load daily training loads from the training_load table (same source as PMC).

    Applies the same cross-modal scaling as compute_and_store_pmc() in
    load_stream.py so that the Banister model is fitted on the identical
    load signal used for CTL/ATL/TSB.

    Returns (daily_loads, min_date_str).
    """
    from datetime import timedelta
    from training_engine.load_stream import _cross_modal_scale

    with conn.cursor() as cur:
        cur.execute("""
            SELECT activity_date, source, load_value
            FROM training_load
            ORDER BY activity_date
        """)
        rows = cur.fetchall()

    if not rows:
        return [], ""

    # Aggregate daily load with cross-modal scaling (mirrors compute_and_store_pmc)
    load_by_date: dict = {}
    for activity_date, source, load_value in rows:
        scale = _cross_modal_scale(source)
        load_by_date[activity_date] = load_by_date.get(activity_date, 0.0) + float(load_value) * scale

    # Fill gaps between first and last date with 0 (rest days)
    start_date = min(load_by_date.keys())
    end_date = max(load_by_date.keys())

    daily_loads = []
    current = start_date
    while current <= end_date:
        day_index = (current - start_date).days
        daily_loads.append((day_index, load_by_date.get(current, 0.0)))
        current += timedelta(days=1)

    return daily_loads, str(start_date)


def fit_from_db(conn, estimated_hrmax: float = 190) -> BanisterParams:
    """End-to-end Banister fitting from database.

    1. Load anchor runs from DB
    2. Load daily training loads from DB
    3. Fit Banister parameters
    4. Store results in banister_params table

    Args:
        conn: psycopg2 database connection.
        estimated_hrmax: Estimated maximum heart rate.

    Returns:
        Fitted BanisterParams.
    """
    # Table 'banister_params' is defined in sync/schema.sql

    # Load data
    anchors = load_anchors_from_db(conn, estimated_hrmax=estimated_hrmax)
    daily_loads, min_date = _load_daily_loads_from_db(conn)

    # Convert anchor dates to day indices relative to the load series
    if min_date and anchors:
        from datetime import date as date_type

        min_d = date_type.fromisoformat(min_date)
        anchor_inputs = []
        for a in anchors:
            anchor_date = date_type.fromisoformat(str(a["date"])[:10])
            day_index = (anchor_date - min_d).days
            anchor_inputs.append({"day_index": day_index, "vdot": a["vdot"]})
    else:
        anchor_inputs = []

    # Fit
    params = fit_banister(daily_loads, anchor_inputs)

    # Store
    with conn.cursor() as cur:
        cur.execute("""
            INSERT INTO banister_params (p0, k1, k2, tau1, tau2, n_anchors, fitted_at)
            VALUES (%s, %s, %s, %s, %s, %s, NOW())
        """, (params.p0, params.k1, params.k2, params.tau1, params.tau2, len(anchor_inputs)))
    conn.commit()

    return params

File: web/lib/banister.ts
/**
 * Banister impulse-response fit — TS port of the DB glue in
 * sync/src/training_engine/banister.py. The model itself (predict + the
 * differential-evolution fit that matches scipy) lives in the `banister` npm
 * package (drkostas/banister, verified against scipy in Phase-0); this module
 * is the soma-specific DB layer: anchor detection from Garmin runs, daily-load
 * loading (same cross-modal signal as the PMC), and fit_from_db which stores
 * banister_params. Stage: training engine (#187). DB-only.
 *
 * The DE fit is stochastic, so fitted params / current_vdot are close to but
 * not bit-identical to the Python (scipy) fit — this is a model fit, not a
 * deterministic transform.
 */
import { banisterPredict, fitBanister, DEFAULT_PARAMS, type BanisterParams, type DailyLoad, type Anchor } from "banister";
import type { QueryFn } from "./db";
import { vdotFromRace } from "./vdot";
import { crossModalScale } from "./pmc-stream";

export type { BanisterParams };
export { banisterPredict, DEFAULT_PARAMS };

/** Today (YYYY-MM-DD) in America/New_York — mirrors config.today_nyc. */
function todayNyc(now: Date = new Date()): string {
  return now.toLocaleDateString("en-CA", { timeZone: "America/New_York" });
}

function daysBetween(a: string, b: string): number {
  return Math.round((Date.parse(b + "T00:00:00Z") - Date.parse(a + "T00:00:00Z")) / 86_400_000);
}

export interface RunInput { date: string; avg_hr: number; distance_m: number; duration_s: number; activity_id?: number; }
export interface AnchorRun extends RunInput { vdot: number; }

/**
 * Detect maximal-effort anchor runs (avg_hr ≥ pct·HRmax AND distance ≥ min),
 * compute VDOT for each, sort by date. Port of detect_anchor_runs.
 */
export function detectAnchorRuns(
  runs: RunInput[], estimatedHrmax: number, hrThresholdPct = 0.9, minDistanceM = 2000,
): AnchorRun[] {
  const hrCutoff = estimatedHrmax * hrThresholdPct;
  const anchors: AnchorRun[] = [];
  for (const run of runs) {
    const avgHr = run.avg_hr || 0, distanceM = run.distance_m || 0, durationS = run.duration_s || 0;
    if (avgHr < hrCutoff || distanceM < minDistanceM || durationS <= 0) continue;
    anchors.push({ ...run, vdot: vdotFromRace(distanceM, durationS) });
  }
  anchors.sort((a, b) => (a.date < b.date ? -1 : a.date > b.date ? 1 : 0));
  return anchors;
}

/** Load running-activity anchors from garmin_activity_raw. Port of load_anchors_from_db. */
export async function loadAnchorsFromDb(sql: QueryFn, estimatedHrmax = 190): Promise<AnchorRun[]> {
  const rows = await sql`
    SELECT activity_id, raw_json FROM garmin_activity_raw
    WHERE endpoint_name = 'summary'
      AND raw_json->'activityType'->>'typeKey' = 'running'
    ORDER BY (raw_json->>'startTimeLocal')::date ASC`;
  const runs: RunInput[] = [];
  for (const row of rows) {
    const d = typeof row.raw_json === "string" ? JSON.parse(row.raw_json) : row.raw_json;
    const avgHr = d.averageHR, distanceM = d.distance, durationS = d.duration, startLocal = d.startTimeLocal || "";
    if (!avgHr || !distanceM || !durationS || !startLocal) continue;
    runs.push({
      date: String(startLocal).slice(0, 10),
      avg_hr: Number(avgHr), distance_m: Number(distanceM), duration_s: Number(durationS),
      activity_id: row.activity_id,
    });
  }
  return detectAnchorRuns(runs, estimatedHrmax);
}

/**
 * Load daily training loads from training_load with the SAME cross-modal scaling
 * as the PMC, gap-filled with 0. Returns [dailyLoads, minDate]. Port of
 * _load_daily_loads_from_db.
 */
export async function loadDailyLoadsFromDb(sql: QueryFn): Promise<[DailyLoad[], string]> {
  const rows = await sql`
    SELECT activity_date::text AS activity_date, source, load_value
    FROM training_load ORDER BY activity_date`;
  if (!rows.length) return [[], ""];

  const loadByDate = new Map<string, number>();
  for (const row of rows) {
    const scale = crossModalScale(row.source);
    loadByDate.set(row.activity_date, (loadByDate.get(row.activity_date) ?? 0) + Number(row.load_value) * scale);
  }
  const keys = [...loadByDate.keys()].sort();
  const startDate = keys[0], endDate = keys[keys.length - 1];
  const dailyLoads: DailyLoad[] = [];
  for (let cur = startDate; cur <= endDate; cur = new Date(Date.parse(cur + "T00:00:00Z") + 86_400_000).toISOString().slice(0, 10)) {
    dailyLoads.push([daysBetween(startDate, cur), loadByDate.get(cur) ?? 0]);
  }
  return [dailyLoads, startDate];
}

export interface BanisterFitResult extends BanisterParams { n_anchors: number; current_vdot: number; }

/**
 * End-to-end Banister fit from DB: load anchors + daily loads, filter anchors to
 * the last 2 years, convert to day indices, fit, predict today's VDOT, and store
 * a banister_params row. Port of fit_from_db. Returns the fitted params +
 * n_anchors + current_vdot. DB.
 */
export async function fitFromDb(sql: QueryFn, estimatedHrmax = 190): Promise<BanisterFitResult> {
  let anchors = await loadAnchorsFromDb(sql, estimatedHrmax);
  const [dailyLoads, minDate] = await loadDailyLoadsFromDb(sql);

  const cutoff = new Date(Date.parse(todayNyc() + "T00:00:00Z") - 730 * 86_400_000).toISOString().slice(0, 10);
  const recent = anchors.filter((a) => a.date.slice(0, 10) >= cutoff);
  if (recent.length >= 2) anchors = recent;

  let anchorInputs: Anchor[] = [];
  if (minDate && anchors.length) {
    anchorInputs = anchors.map((a) => ({ day_index: daysBetween(minDate, a.date.slice(0, 10)), vdot: a.vdot }));
  }

  const params = fitBanister(dailyLoads, anchorInputs);
  const todayIdx = minDate ? daysBetween(minDate, todayNyc()) : 0;
  const currentVdot = dailyLoads.length ? banisterPredict(params, dailyLoads, todayIdx) : params.p0;

  await sql`
    INSERT INTO banister_params (p0, k1, k2, tau1, tau2, n_anchors, current_vdot, fitted_at)
    VALUES (${params.p0}, ${params.k1}, ${params.k2}, ${params.tau1}, ${params.tau2},
            ${anchorInputs.length}, ${currentVdot}, NOW())`;

  return { ...params, n_anchors: anchorInputs.length, current_vdot: currentVdot };
}
